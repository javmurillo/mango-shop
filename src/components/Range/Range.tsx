import React, { Component } from 'react';
import styled from 'styled-components';
import RangeInput from '../RangeInput/RangeInput';
import RangeSlider from '../RangeSlider/RangeSlider';
import { RangeProps } from './range-props.interface';
import { RangeState } from './range-state.interface';

const StyledMainRange = styled.div`
  width: 512px;
  text-align: center;
  display: flex;
  flex-direction: row;
  align-items: center;
  margin-top: 16px;
`;

const StyledWrapperRange = styled.div`
  width: 75%;
`;

export default class Range extends Component<RangeProps, RangeState> {
  public state: RangeState = {
    rangeValue: {
      start: 0,
      end: 100,
    },
    min: 0,
    max: 100,
  };

  constructor(props: RangeProps) {
    super(props);
    const { min, max, step } = props;
    if (Array.isArray(step)) {
      const stepArray = step as number[];
      const minArray = stepArray[0];
      const maxArray = stepArray[stepArray.length - 1];
      this.state = {
        min: minArray,
        max: maxArray,
        rangeValue: {
          start: minArray,
          end: maxArray,
        },
      };
    } else {
      this.state = {
        min,
        max,
        rangeValue: {
          start: min,
          end: max,
        },
      };
    }
  }

  private onChange = (rangeValue: { start: number; end: number }): void => {
    this.setState({ rangeValue });
    this.props.onFilterArticles(rangeValue.start, rangeValue.end);
  };

  private handleChange = (
    event: React.ChangeEvent<HTMLInputElement>,
    key: 'start' | 'end'
  ): void => {
    const target = event.target;
    const newValue = Number(target.value);
    const { min, max, rangeValue } = this.state;
    if (key === 'start') {
      const canMoveStart =
        newValue >= 0 && newValue >= min && newValue <= rangeValue.end;
      if (canMoveStart) {
        this.patchState(key, newValue);
      }
    } else {
      const canMoveEnd =
        newValue && newValue <= max && newValue >= rangeValue.start;
      if (canMoveEnd) {
        this.patchState(key, newValue);
      }
    }
  };

  private patchState(key: 'start' | 'end', value: number): void {
    this.setState(
      previousState => ({
        ...previousState,
        rangeValue: {
          ...previousState.rangeValue,
          [key]: value,
        },
      }),
      () => {
        this.props.onFilterArticles(
          this.state.rangeValue.start,
          this.state.rangeValue.end
        );
      }
    );
  }

  render(): JSX.Element {
    return (
      <StyledMainRange>
        <RangeInput
          onChange={this.handleChange}
          value={this.state.rangeValue.start}
          rangeKey="start"
          disabled={this.props.disableInputs}
        ></RangeInput>

        <RangeSlider
          min={this.state.min}
          max={this.state.max}
          step={this.props.step}
          rangeValue={this.state.rangeValue}
          onChange={this.onChange}
        />

        <RangeInput
          onChange={this.handleChange}
          value={this.state.rangeValue.end}
          rangeKey="end"
          disabled={this.props.disableInputs}
        ></RangeInput>
      </StyledMainRange>
    );
  }
}
