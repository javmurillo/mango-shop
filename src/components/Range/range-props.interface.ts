export interface RangeProps {
  onFilterArticles: (min: number, max: number) => void;
  min: number;
  max: number;
  step: number | number[];
  disableInputs: boolean;
}
