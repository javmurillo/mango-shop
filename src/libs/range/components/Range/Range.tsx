import React, { Component } from 'react';
import styled from 'styled-components';
import RangeInput from '../RangeInput/RangeInput';
import RangeSlider from '../RangeSlider/RangeSlider';
import { RangeProps } from './range-props.interface';
import { RangeState } from './range-state.interface';

const StyledMainRange = styled.div`
  width: 512px;
  text-align: center;
  display: flex;
  flex-direction: row;
  align-items: center;
  margin-top: 16px;
`;

/**
 * Range class
 */
export default class Range extends Component<RangeProps, RangeState> {
  public state: RangeState; // Range state

  constructor(props: RangeProps) {
    super(props);
    let { min, max, step } = props;
    if (Array.isArray(step)) {
      step.sort((a, b) => a - b);
      if (step.length > 0) {
        min = step[0];
        max = step[step.length - 1];
      } else {
        step = 1;
      }
    } else {
      max = Math.max(props.min, props.max);
      min = Math.min(props.min, props.max);
    }
    // Initializing state
    this.state = {
      step,
      min,
      max,
      rangeValue: {
        start: min,
        end: max,
      },
    };
    console.log(this.state);
  }

  /**
   * Updates the state and calls the [onChange] function.
   * @param rangeValue Rage of the current values
   */
  private onChange = (rangeValue: { start: number; end: number }): void => {
    this.setState({ rangeValue });
    if (this.props.onChange) {
      this.props.onChange(rangeValue.start, rangeValue.end);
    }
  };

  /**
   * Updates the state given the value from the input and the [key] value: starting or ending input.
   * @param event React.ChangeEvent<HTMLInputElement>
   * @param key Starting or Ending input.
   */
  private handleInputChange = (
    event: React.ChangeEvent<HTMLInputElement>,
    key: 'start' | 'end'
  ): void => {
    const target = event.target;
    const newValue = Number(target.value);
    const { min, max, rangeValue } = this.state;
    if (key === 'start') {
      if (!newValue) {
        this.patchState(key, min);
      } else if (newValue >= min && newValue <= rangeValue.end) {
        this.patchState(key, newValue);
      }
    } else if (key === 'end') {
      if (!newValue) {
        this.patchState(key, max);
      } else if (newValue <= max && newValue >= rangeValue.start) {
        this.patchState(key, newValue);
      }
    }
  };

  /**
   * Updates the state and calls the [onChange] function once finished.
   * @param event React.ChangeEvent<HTMLInputElement>
   * @param key Starting or Ending input.
   */
  private patchState = (key: 'start' | 'end', value: number): void => {
    this.setState(
      previousState => ({
        ...previousState,
        rangeValue: {
          ...previousState.rangeValue,
          [key]: value,
        },
      }),
      () => {
        if (this.props.onChange) {
          this.props.onChange(
            this.state.rangeValue.start,
            this.state.rangeValue.end
          );
        }
      }
    );
  };

  render(): JSX.Element {
    return (
      <StyledMainRange aria-label={this.props.ariaLabel}>
        <RangeInput
          onChange={this.handleInputChange}
          value={this.state.rangeValue.start}
          rangeKey="start"
          disabled={this.props.disableInputs}
        ></RangeInput>
        <RangeSlider
          min={this.state.min}
          max={this.state.max}
          step={this.state.step}
          rangeValue={this.state.rangeValue}
          onChange={this.onChange}
          ariaLabel="range-slider"
        />
        <RangeInput
          onChange={this.handleInputChange}
          value={this.state.rangeValue.end}
          rangeKey="end"
          disabled={this.props.disableInputs}
        ></RangeInput>
      </StyledMainRange>
    );
  }
}
