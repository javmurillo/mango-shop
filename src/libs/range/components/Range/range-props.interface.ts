export interface RangeProps {
  onChange: (min: number, max: number) => void;
  min: number;
  max: number;
  step: number | number[];
  disableInputs: boolean;
}
