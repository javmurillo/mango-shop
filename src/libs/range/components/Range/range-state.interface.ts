export interface RangeState {
  rangeValue: {
    start: number;
    end: number;
  };
  min: number;
  max: number;
  step: number | number[];
}
