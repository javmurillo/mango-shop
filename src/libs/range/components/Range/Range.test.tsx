import { cleanup, render } from '@testing-library/react';
import TestRenderer from 'react-test-renderer';
import Range from './Range';

afterEach(cleanup);

const setup = () => {
  const utils = render(<Range min={0} max={100} ariaLabel="range" step={5} />);
  const range = utils.getByLabelText('range');
  return {
    range,
    ...utils,
  };
};

describe('<RangeBullet/> tests', () => {
  test('renders correctly', () => {
    const { range } = setup();

    expect(range).toBeInTheDocument();
  });

  test('state initialized correctly', () => {
    const { range } = setup();

    expect(range).toBeInTheDocument();
  });

  test('shoud match snapshot', () => {
    const testRendererJson = TestRenderer.create(
      <Range min={0} max={100} ariaLabel="range" step={5} />
    ).toJSON();
    expect(testRendererJson).toMatchSnapshot();
  });
});
