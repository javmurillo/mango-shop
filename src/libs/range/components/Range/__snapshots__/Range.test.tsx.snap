// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RangeBullet/> tests shoud match snapshot 1`] = `
<div
  aria-label="range"
  className="sc-jSgupP gpqVZN"
>
  <div
    className="sc-gsTCUz ywWHk"
  >
    <input
      className="sc-bdfBwQ iwQBvM"
      min="0"
      onChange={[Function]}
      style={Object {}}
      type="number"
      value="0"
    />
    <span
      className="sc-dlfnbm jjMaay"
    >
      €
    </span>
  </div>
  <div
    aria-label="range-slider"
    className="sc-eCssSg bQcZvo"
  >
    <div
      className="sc-hKgILt mbicC"
    />
    <div
      onMouseDown={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      style={
        Object {
          "backgroundColor": "black",
          "border": "1px solid #052350",
          "borderRadius": "50%",
          "cursor": "grab",
          "display": "inline-block",
          "height": 30,
          "left": "0%",
          "position": "absolute",
          "width": 30,
          "zIndex": 1,
        }
      }
    />
    <div
      onMouseDown={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      style={
        Object {
          "backgroundColor": "black",
          "border": "1px solid #052350",
          "borderRadius": "50%",
          "cursor": "grab",
          "display": "inline-block",
          "height": 30,
          "left": "0%",
          "position": "absolute",
          "width": 30,
          "zIndex": 1,
        }
      }
    />
  </div>
  <div
    className="sc-gsTCUz ywWHk"
  >
    <input
      className="sc-bdfBwQ iwQBvM"
      min="0"
      onChange={[Function]}
      style={Object {}}
      type="number"
      value="100"
    />
    <span
      className="sc-dlfnbm jjMaay"
    >
      €
    </span>
  </div>
</div>
`;
